File: Sources/JVoice/Services/Transcription/BenchRunner.swift
import Foundation

/// Hidden CLI bench mode:
///
///     JVoice --bench <audio.wav> [--model tiny|base|small|large] [--vocab "Word1,Word2"] [--stream [--realtime]] [--repeat N [--idle S]]
///
/// Transcribes one file with timing and prints both the raw transcript and the
/// TextProcessor-processed output. Dev-only: this machine cannot execute
/// XCTest, so this is how transcription speed and vocabulary biasing are
/// actually verified end-to-end.
enum BenchRunner {
    static func shouldRun(arguments: [String]) -> Bool {
        arguments.contains("--bench")
    }

    static func runAndExit(arguments: [String]) -> Never {
        var exitCode: Int32 = 0
        let semaphore = DispatchSemaphore(value: 0)
        Task.detached {
            exitCode = await run(arguments: arguments)
            semaphore.signal()
        }
        semaphore.wait()
        exit(exitCode)
    }

    private static func run(arguments: [String]) async -> Int32 {
        guard let benchIndex = arguments.firstIndex(of: "--bench"),
              arguments.count > benchIndex + 1 else {
            FileHandle.standardError.write(Data("usage: JVoice --bench <audio.wav> [--model tiny|base|small|large] [--lang en|ro] [--vocab \"Word1,Word2\"] [--stream [--realtime]] [--repeat N [--idle S]]\n".utf8))
            return 64
        }
        let audioURL = URL(fileURLWithPath: arguments[benchIndex + 1])
        guard FileManager.default.fileExists(atPath: audioURL.path) else {
            FileHandle.standardError.write(Data("no such file: \(audioURL.path)\n".utf8))
            return 66
        }

        var model: WhisperModelOption = .base
        if let modelIndex = arguments.firstIndex(of: "--model"), arguments.count > modelIndex + 1 {
            switch arguments[modelIndex + 1] {
            case "tiny": model = .tiny
            case "base": model = .base
            case "small": model = .small
            case "large", "large-v3_turbo", "largeTurbo", "large-v3-v20240930": model = .largeTurbo
            default:
                FileHandle.standardError.write(Data("unknown model \(arguments[modelIndex + 1])\n".utf8))
                return 64
            }
        }

        var language: TranscriptionLanguage = .english
        if let langIndex = arguments.firstIndex(of: "--lang"), arguments.count > langIndex + 1 {
            switch arguments[langIndex + 1] {
            case "en", "english": language = .english
            case "ro", "romanian": language = .romanian
            default:
                FileHandle.standardError.write(Data("unknown lang \(arguments[langIndex + 1])\n".utf8))
                return 64
            }
        }

        var vocabulary: [String] = []
        if let vocabIndex = arguments.firstIndex(of: "--vocab"), arguments.count > vocabIndex + 1 {
            vocabulary = arguments[vocabIndex + 1]
                .split(separator: ",")
                .map { $0.trimmingCharacters(in: .whitespaces) }
                .filter { !$0.isEmpty }
        }

        let useDecoderPrompt = !arguments.contains("--no-prompt")
        print("model: \(model.rawValue)   audio: \(audioURL.lastPathComponent)   lang: \(language.whisperCode)   vocab: \(vocabulary.isEmpty ? "—" : vocabulary.joined(separator: ", "))   decoderPrompt: \(useDecoderPrompt ? "on" : "off")")

        #if canImport(WhisperKit)
        let engine = WhisperKitTranscriptionEngine(model: model, language: language, vocabulary: vocabulary, useVocabularyPrompt: useDecoderPrompt)

        let loadStart = Date()
        await engine.prewarm()
        print(String(format: "load+prewarm: %.2fs", Date().timeIntervalSince(loadStart)))

        if arguments.contains("--stream") {
            return await runStream(audioURL: audioURL, engine: engine, realtime: arguments.contains("--realtime"))
        }

        do {
            // --repeat N [--idle S]: decode the same clip N times in ONE process,
            // sleeping S seconds between runs — separates cold (first decode after
            // idle: Neural Engine/CPU power state) from warm per-decode cost.
            if let repeatIndex = arguments.firstIndex(of: "--repeat"), arguments.count > repeatIndex + 1,
               let repeats = Int(arguments[repeatIndex + 1]), repeats > 1 {
                var idle: Double = 0
                if let idleIndex = arguments.firstIndex(of: "--idle"), arguments.count > idleIndex + 1 {
                    idle = Double(arguments[idleIndex + 1]) ?? 0
                }
                for i in 1...repeats {
                    if i > 1, idle > 0 { try? await Task.sleep(nanoseconds: UInt64(idle * 1_000_000_000)) }
                    let t0 = Date()
                    _ = try? await engine.transcribe(audioURL: audioURL)
                    print(String(format: "run %d%@: %.2fs", i, idle > 0 && i > 1 ? " (after \(Int(idle)) s idle)" : "", Date().timeIntervalSince(t0)))
                }
                return 0
            }
            let transcribeStart = Date()
            let raw = try await engine.transcribe(audioURL: audioURL)
            let elapsed = Date().timeIntervalSince(transcribeStart)
            print(String(format: "transcribe:   %.2fs", elapsed))
            print("raw:       \"\(raw)\"")
            let userDict = TextProcessor.buildUserDictionary(from: vocabulary)
            let processed = TextProcessor.process(raw, mode: .casual, extraDictionary: userDict, vocabulary: vocabulary)
            print("processed: \"\(processed)\"")
            return 0
        } catch {
            FileHandle.standardError.write(Data("transcription failed: \(error.localizedDescription)\n".utf8))
            return 1
        }
        #else
        FileHandle.standardError.write(Data("WhisperKit unavailable in this build\n".utf8))
        return 70
        #endif
    }

    #if canImport(WhisperKit)
    /// Streaming E2E without a microphone: replays `audioURL` into a growing
    /// temp WAV at ~10× real time (or 1× with `--realtime`, which also uses the
    /// app's own poll cadence + speculative tail decode so a clip that ends in
    /// a pause shows the "speculative tail HIT" path) while a real
    /// StreamingTranscriptionSession consumes it, then compares against the
    /// whole-file transcript.
    private static func runStream(audioURL: URL, engine: WhisperKitTranscriptionEngine, realtime: Bool) async -> Int32 {
        guard let sourceBytes = try? Data(contentsOf: audioURL),
              let info = WavTail.parseHeader([UInt8](sourceBytes.prefix(WavTail.headerProbeBytes))) else {
            FileHandle.standardError.write(Data("not a 16 kHz mono 16-bit PCM wav: \(audioURL.path)\n".utf8))
            return 65
        }
        let pollNanoseconds: UInt64 = realtime ? UInt64(AppTimings.streamingPoll * 1_000_000_000) : 100_000_000
        guard let session = await engine.makeStreamingSession(pollNanoseconds: pollNanoseconds,
                                                              log: { print("  [stream] \($0)") }) else {
            FileHandle.standardError.write(Data("engine has no loaded model\n".utf8))
            return 70
        }

        let growingURL = FileManager.default.temporaryDirectory
            .appendingPathComponent("jv-stream-\(UUID().uuidString).wav")
        defer { try? FileManager.default.removeItem(at: growingURL) }
        FileManager.default.createFile(atPath: growingURL.path, contents: sourceBytes.prefix(info.dataOffset))

        let payload = sourceBytes.dropFirst(info.dataOffset)
        let sliceBytes = info.sampleRate * info.bytesPerSample / 2 // 0.5 s of audio…
        let writer = Task {
            guard let handle = try? FileHandle(forWritingTo: growingURL) else { return }
            defer { try? handle.close() }
            _ = try? handle.seekToEnd()
            var offset = payload.startIndex
            while offset < payload.endIndex {
                let end = min(offset + sliceBytes, payload.endIndex)
                try? handle.write(contentsOf: payload[offset..<end])
                offset = end
                // …every 50 ms ⇒ 10× real time; every 500 ms ⇒ real time
                try? await Task.sleep(nanoseconds: realtime ? 500_000_000 : 50_000_000)
            }
        }

        let wallStart = Date()
        await session.start(url: growingURL)
        await writer.value // "recording" ends here
        let stopStart = Date()
        let streamed = await session.finish()
        let tailTime = Date().timeIntervalSince(stopStart)
        let wallTime = Date().timeIntervalSince(wallStart)

        print(String(format: "stream wall: %.2fs   post-stop (finish): %.2fs", wallTime, tailTime))
        print("streamed:  \(streamed.map { "\"\($0)\"" } ?? "nil (session fell back)")")

        do {
            let wholeStart = Date()
            let whole = try await engine.transcribe(audioURL: audioURL)
            print(String(format: "wholefile: %.2fs", Date().timeIntervalSince(wholeStart)))
            print("wholefile: \"\(whole)\"")
        } catch {
            FileHandle.standardError.write(Data("whole-file comparison failed: \(error.localizedDescription)\n".utf8))
            return 1
        }
        return 0
    }
    #endif
}

File: windows/JVoice.App/Whisper/BenchRunner.cs
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using JVoice.Core.Audio;
using JVoice.Core.Models;
using JVoice.Core.Text;
using Whisper.net.LibraryLoader;

namespace JVoice.App.Whisper;

/// Hidden CLI bench mode (port of BenchRunner.swift):
///
///     JVoice --bench &lt;audio.wav&gt; [--model tiny|base|small|large] [--lang en|ro]
///            [--vocab "Word1,Word2"] [--stream] [--no-prompt]
///
/// Transcribes one file with timing and prints the raw transcript and the
/// TextProcessor-processed output. The primary end-to-end verification of
/// transcription speed and vocabulary biasing on Windows (no XUnit coverage of
/// the native path). Runs BEFORE any UI (Program.Main / Phase 4 App startup).
internal static class BenchRunner
{
    public static bool ShouldRun(string[] arguments) => arguments.Contains("--bench");

    /// Blocks on the async run and returns the process exit code.
    public static int RunAndExit(string[] arguments)
        => RunAsync(arguments).GetAwaiter().GetResult();

    private static async Task<int> RunAsync(string[] arguments)
    {
        int benchIndex = Array.IndexOf(arguments, "--bench");
        if (benchIndex < 0 || arguments.Length <= benchIndex + 1)
        {
            Console.Error.WriteLine(
                "usage: JVoice --bench <audio.wav> [--model tiny|base|small|large] [--lang en|ro] " +
                "[--vocab \"Word1,Word2\"] [--stream] [--no-prompt] " +
                "[--iters N] [--flash on|off] [--threads N] [--audio-ctx off|auto|N] " +
                "[--runtime auto|cuda|cuda12|cuda-any|vulkan|cpu] [--log-runtime]");
            return 64;
        }

        string audioPath = arguments[benchIndex + 1];
        if (!File.Exists(audioPath))
        {
            Console.Error.WriteLine($"no such file: {audioPath}");
            return 66;
        }

        var model = WhisperModelOption.Base;
        int modelIndex = Array.IndexOf(arguments, "--model");
        if (modelIndex >= 0 && arguments.Length > modelIndex + 1)
        {
            switch (arguments[modelIndex + 1])
            {
                case "tiny": model = WhisperModelOption.Tiny; break;
                case "base": model = WhisperModelOption.Base; break;
                case "small": model = WhisperModelOption.Small; break;
                case "large":
                case "large-v3_turbo":
                case "largeTurbo":
                case "large-v3-v20240930":
                    model = WhisperModelOption.LargeTurbo; break;
                default:
                    Console.Error.WriteLine($"unknown model {arguments[modelIndex + 1]}");
                    return 64;
            }
        }

        var language = TranscriptionLanguage.English;
        int langIndex = Array.IndexOf(arguments, "--lang");
        if (langIndex >= 0 && arguments.Length > langIndex + 1)
        {
            switch (arguments[langIndex + 1])
            {
                case "en":
                case "english": language = TranscriptionLanguage.English; break;
                case "ro":
                case "romanian": language = TranscriptionLanguage.Romanian; break;
                default:
                    Console.Error.WriteLine($"unknown lang {arguments[langIndex + 1]}");
                    return 64;
            }
        }

        var vocabulary = new List<string>();
        int vocabIndex = Array.IndexOf(arguments, "--vocab");
        if (vocabIndex >= 0 && arguments.Length > vocabIndex + 1)
        {
            vocabulary = arguments[vocabIndex + 1]
                .Split(',')
                .Select(s => s.Trim())
                .Where(s => s.Length > 0)
                .ToList();
        }

        bool useDecoderPrompt = !arguments.Contains("--no-prompt");

        // ---- speed-measurement flags (speed plan 2026-06-27) ----
        int iters = ReadIntFlag(arguments, "--iters", 3);

        bool? flash = null;
        int fi = Array.IndexOf(arguments, "--flash");
        if (fi >= 0 && arguments.Length > fi + 1)
            flash = arguments[fi + 1] is "on" or "true" or "1";

        int? threads = null;
        int thi = Array.IndexOf(arguments, "--threads");
        if (thi >= 0 && arguments.Length > thi + 1 && int.TryParse(arguments[thi + 1], out int tparsed))
            threads = tparsed;

        // --audio-ctx: "off" (default, full window) | "auto" (per-clip policy) | <int> (fixed)
        bool adaptiveCtx = false;
        int? fixedCtx = null;
        int aci = Array.IndexOf(arguments, "--audio-ctx");
        if (aci >= 0 && arguments.Length > aci + 1)
        {
            string v = arguments[aci + 1];
            if (v == "auto") adaptiveCtx = true;
            else if (v != "off" && int.TryParse(v, out int cparsed)) fixedCtx = cparsed;
        }

        if (arguments.Contains("--log-runtime")) WhisperRuntime.EnableDebugLogging();

        int ri = Array.IndexOf(arguments, "--runtime");
        if (ri >= 0 && arguments.Length > ri + 1)
        {
            switch (arguments[ri + 1])
            {
                case "auto": break; // default probe order
                case "cuda": WhisperRuntime.ForceRuntimeOrder(RuntimeLibrary.Cuda); break;
                case "cuda12": WhisperRuntime.ForceRuntimeOrder(RuntimeLibrary.Cuda12); break;
                case "cuda-any": WhisperRuntime.ForceRuntimeOrder(RuntimeLibrary.Cuda, RuntimeLibrary.Cuda12); break;
                case "vulkan": WhisperRuntime.ForceRuntimeOrder(RuntimeLibrary.Vulkan); break;
                case "cpu": WhisperRuntime.ForceRuntimeOrder(RuntimeLibrary.Cpu); break;
                default:
                    Console.Error.WriteLine($"unknown runtime {arguments[ri + 1]}");
                    return 64;
            }
        }

        var tuning = new EngineTuning(
            UseFlashAttention: flash ?? false,
            AdaptiveAudioContext: adaptiveCtx,
            FixedAudioContext: fixedCtx,
            Threads: threads);

        // Matches the Swift header line. Swift printed `model.rawValue`; we print the
        // GGML file stem (the closest stable Windows analog of the model identity).
        string vocabDisplay = vocabulary.Count == 0 ? "—" : string.Join(", ", vocabulary);
        Console.WriteLine(
            $"model: {model.GgmlFileName()}   audio: {Path.GetFileName(audioPath)}   " +
            $"lang: {language.WhisperCode()}   vocab: {vocabDisplay}   " +
            $"decoderPrompt: {(useDecoderPrompt ? "on" : "off")}");
        Console.WriteLine(
            $"tuning: flash={(tuning.UseFlashAttention ? "on" : "off")}  " +
            $"threads={(tuning.Threads?.ToString() ?? "default")}  " +
            $"audio_ctx={(tuning.FixedAudioContext?.ToString() ?? (tuning.AdaptiveAudioContext ? "auto" : "full"))}  " +
            $"iters={iters}");

        var store = new WhisperModelStore();
        var engine = new WhisperNetTranscriptionEngine(
            model, language, vocabulary, useDecoderPrompt, store, tuning);

        var loadSw = Stopwatch.StartNew();
        await engine.PrewarmAsync();
        loadSw.Stop();
        WhisperRuntime.EnsureLoaded();
        Console.WriteLine($"runtime: {WhisperRuntime.Describe()}");
        Console.WriteLine($"load+prewarm: {loadSw.Elapsed.TotalSeconds:0.00}s");

        if (!await engine.IsReadyAsync())
        {
            Console.Error.WriteLine("engine has no loaded model (download or runtime failure)");
            return 70;
        }

        if (arguments.Contains("--stream"))
            return await RunStreamAsync(audioPath, engine, vocabulary);

        try
        {
            // One warm-up decode (excluded) so steady-state timing isn't polluted by lazy init.
            string raw = await engine.TranscribeAsync(audioPath, CancellationToken.None);

            var times = new List<double>();
            for (int i = 0; i < Math.Max(1, iters); i++)
            {
                var sw = Stopwatch.StartNew();
                raw = await engine.TranscribeAsync(audioPath, CancellationToken.None);
                sw.Stop();
                times.Add(sw.Elapsed.TotalSeconds);
            }
            times.Sort();
            double median = times[times.Count / 2];
            Console.WriteLine(
                $"transcribe: median={median:0.000}s  min={times[0]:0.000}s  " +
                $"max={times[^1]:0.000}s  (n={times.Count}, warm-up excluded)");
            Console.WriteLine($"raw:       \"{raw}\"");
            var userDict = TextProcessor.BuildUserDictionary(vocabulary);
            string processed = TextProcessor.Process(
                raw, ToneStyle.Casual, userDict, removeFillerWords: false, vocabulary: vocabulary);
            Console.WriteLine($"processed: \"{processed}\"");
            return 0;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"transcription failed: {ex.Message}");
            return 1;
        }
    }

    private static int ReadIntFlag(string[] args, string name, int fallback)
    {
        int i = Array.IndexOf(args, name);
        return (i >= 0 && args.Length > i + 1 && int.TryParse(args[i + 1], out int v)) ? v : fallback;
    }

    /// Streaming E2E without a microphone: replays `audioPath` into a growing temp
    /// WAV at ~10× real time while a real StreamingTranscriptionSession consumes it,
    /// then compares against the whole-file transcript. Port of BenchRunner.runStream.
    private static async Task<int> RunStreamAsync(
        string audioPath, WhisperNetTranscriptionEngine engine, IReadOnlyList<string> vocabulary)
    {
        byte[] sourceBytes;
        try { sourceBytes = await File.ReadAllBytesAsync(audioPath); }
        catch (Exception)
        {
            Console.Error.WriteLine($"cannot read file: {audioPath}");
            return 65;
        }

        int probe = Math.Min(sourceBytes.Length, WavTail.HeaderProbeBytes);
        var info = WavTail.ParseHeader(sourceBytes.AsSpan(0, probe));
        if (info is not { } header)
        {
            Console.Error.WriteLine($"not a 16 kHz mono 16-bit PCM wav: {audioPath}");
            return 65;
        }

        var session = engine.MakeStreamingSession(pollMilliseconds: 100);
        if (session is null)
        {
            Console.Error.WriteLine("engine has no loaded model");
            return 70;
        }

        string growingPath = Path.Combine(
            Path.GetTempPath(), $"jv-stream-{Guid.NewGuid():N}.wav");
        try
        {
            // Seed the growing file with just the header (everything up to dataOffset).
            await File.WriteAllBytesAsync(growingPath, sourceBytes[..header.DataOffset]);

            int sliceBytes = header.SampleRate * header.BytesPerSample / 2; // 0.5 s of audio
            var writer = Task.Run(async () =>
            {
                await using var handle = new FileStream(
                    growingPath, FileMode.Open, FileAccess.Write, FileShare.ReadWrite);
                handle.Seek(0, SeekOrigin.End);
                int offset = header.DataOffset;
                while (offset < sourceBytes.Length)
                {
                    int end = Math.Min(offset + sliceBytes, sourceBytes.Length);
                    await handle.WriteAsync(sourceBytes.AsMemory(offset, end - offset));
                    await handle.FlushAsync();
                    offset = end;
                    await Task.Delay(50); // …every 50 ms ⇒ ~10× real time
                }
            });

            var wallSw = Stopwatch.StartNew();
            session.Start(growingPath);
            await writer;                       // "recording" ends here
            var stopSw = Stopwatch.StartNew();
            string? streamed = await session.Finish();
            stopSw.Stop();
            wallSw.Stop();

            Console.WriteLine(
                $"stream wall: {wallSw.Elapsed.TotalSeconds:0.00}s   " +
                $"post-stop (finish): {stopSw.Elapsed.TotalSeconds:0.00}s");
            Console.WriteLine($"streamed:  {(streamed is null ? "nil (session fell back)" : $"\"{streamed}\"")}");

            try
            {
                var wholeSw = Stopwatch.StartNew();
                string whole = await engine.TranscribeAsync(audioPath, CancellationToken.None);
                wholeSw.Stop();
                Console.WriteLine($"wholefile: {wholeSw.Elapsed.TotalSeconds:0.00}s");
                Console.WriteLine($"wholefile: \"{whole}\"");
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"whole-file comparison failed: {ex.Message}");
                return 1;
            }
            return 0;
        }
        finally
        {
            try { if (File.Exists(growingPath)) File.Delete(growingPath); } catch (IOException) { }
        }
    }
}
